src/rsz/src/GateCloner.cc

```diff
@@ -0,0 +1,763 @@
+  SteinerPt right = tree->right(current);
+  if (left != SteinerNull) {
+    float cap_left = resizer_->subtreeLoad(tree, cap_per_micron, left) + src_wire_cap;
+    bool is_leaf
+        = tree->left(left) == SteinerNull && tree->right(left) == SteinerNull;
+    if (cap_left < c_limit || is_leaf) {
+      cloneInstance(tree, left, current, driver_cell);
+    } else {
+      topDownClone(tree, left, current, c_limit, driver_cell);
+    }
+  }
+
+  if (right != SteinerNull) {
+    float cap_right = resizer_->subtreeLoad(tree, cap_per_micron, right) + src_wire_cap;
+    bool is_leaf
+        = tree->left(right) == SteinerNull && tree->right(right) == SteinerNull;
+    if (cap_right < c_limit || is_leaf) {
+      cloneInstance(tree, right, current, driver_cell);
+    } else {
+      topDownClone(tree, right, current, c_limit, driver_cell);
+    }
+  }
+}
+
+void GateCloner::topDownConnect(SteinerTree *tree, SteinerPt current, Net* net)
+{
+  if (current == SteinerNull) {
+    return;
+  }
+  if (tree->left(current) == SteinerNull && tree->right(current) == SteinerNull) {
+    Instance *inst      = network_->instance(tree->pin(current));
+    auto term_port = network_->port(tree->pin(current));
+    sta_->connectPin(inst, term_port, net);
+  }
+  else {
+    topDownConnect(tree, tree->left(current), net);
+    topDownConnect(tree, tree->right(current), net);
+  }
+}
+void GateCloner::cloneInstance(SteinerTree *tree, SteinerPt current, SteinerPt prev,
+                               LibertyCell* driver_cell)
+{
+  SteinerPt drvr = tree->drvrPt();
+  const Pin *output_pin = tree->pin(drvr);
+  auto inst = network_->instance(output_pin);
+  Net* output_net = network_->net(output_pin);
+
+  // TODO remove this if condition later
+  if  (output_net == nullptr) {
+    return;
+  }
+  resizer_->estimateWireParasitic(output_net);
+  sta_->ensureLevelized();
+  sta_->findRequireds();
+  sta_->findDelays();
+
+  auto wp = worstSlackPath(output_pin);
+  if (wp.empty()) {
+    return;
+  }
+
+  Net* clone_net = resizer_->Resizer::makeUniqueNet();
+  LibertyPort *output_port = network_->libertyPort(output_pin);
+  // Double check what this topDownConnect fun does.
+  topDownConnect(tree, current, clone_net);
+
+  std::unordered_set<Net*> para_nets;
+  para_nets.insert(clone_net);
+  para_nets.insert(output_net);
+  resizer_->estimateWireParasitic(clone_net);
+
+  int fanout_count = fanoutPins(network_->net(output_pin)).size();
+  if (fanout_count == 0) {
+    // Disconnect all the pins
+    for (auto& pin : resizer_->getPins(clone_net)) {
+      sta_->disconnectPin(const_cast<Pin*>(pin));
+    }
+    // now connect them to the right net.
+    topDownConnect(tree, current, output_net);
+    sta_->deleteNet(clone_net);
+    para_nets.erase(clone_net);
+  }
+  else {
+    if (!wp.empty()) {
+      std::string instance_name =  resizer_->makeUniqueInstName("clone_");
+      // Make a cloned instance and connect it to the design
+      Instance *cloned_inst =
```

Comment:
How is this new instance being placed?

Reply:
Good catch. What happens if I don't provide a placement location? Will fix.

Reply:
I would have expected dpl to error or at least warn.  I think it just gets the location blindly which will be (0,0) if unplaced.